internal/util/observability/funccall.go

```diff
@@ -53,14 +55,19 @@ func FuncCall(ctx context.Context) func() {
 		f, _ := runtime.CallersFrames(pc).Next()
 		funcName := f.Function
 
-		fc.region = trace.StartRegion(ctx, funcName)
+		fc.region = trace.StartRegion(parent, funcName)
 	}
 
-	return fc.leave
+	ctx, cancel = context.WithCancel(parent)
+	go fc.run(ctx)
+
+	return
 }
 
-// leave is called on function exit.
-func (fc *funcCall) leave() {
+// run waits until ctx is canceled.
+func (fc *funcCall) run(ctx context.Context) {
```

Comment:
I don't really like the "run" name because, in fact, this function doesn't run anything, it rather waits. I called it "run" to keep it consistent with other similar functions, but better naming proposals are welcome.